src/util/prolog/prologStringManipulation.ts

```diff
@@ -1,41 +1,122 @@
-const fetchEmptyLinesRegExp = /\n(?:\s)*\n/gm;
-const fetchDoubleWhitespacesRegExp = /  */gm;
+import {fetchAllMatchesForAGroup} from "../regExpUtils";
+import builtInPredicates from "./builtInPredicates";
 
 export function removeFileExtension(path: string) {
+	const fetchFileExtension = /(.*)\.[^\.]*/;
+
 	const fetch: string[] | null = fetchFileExtension.exec(path);
 	if (fetch === null) { return ""; }
 	return fetch[1];
 }
 
 export function removeModuleDeclarations(prologProgram: string) {
+	const fetchModuleDeclaration = /:-( |)module\([^)]*\) *\.[\r\t\f\v ]*\n?/g;
+
 	return prologProgram.replace(fetchModuleDeclaration, "");
 }
 
 function removeFullLineComments(prologProgram: string) {
+	const fetchFullLineCommentsRegExp = /^[\r\t\f\v ]*%[^\n]*\n/gm;
+
 	return prologProgram.replace(fetchFullLineCommentsRegExp, "");
 }
 
 function removePartialLineComments(prologProgram: string) {
+	const fetchPartialLineCommentsRegExp = /(^[^\n%]+)%[^\n\r]*(\r?\n)/gm;
+
 	// group2 fetches the EOL standard in the file to fit both: Windows and UNIX
 	return prologProgram.replace(fetchPartialLineCommentsRegExp, (_, group1, group2) => group1.trimRight() + group2);
 }
 
 export function removeComments(prologProgram: string) {
-	return removeFullLineComments(removePartialLineComments(prologProgram));
+
+ return removeFullLineComments(removePartialLineComments(prologProgram));
 }
 
 function removeEmptyLines(prologProgram: string) {
+	const fetchEmptyLinesRegExp = /\n(?:\s)*\n/gm;
+
 	return prologProgram.replace(fetchEmptyLinesRegExp, "\n");
 }
 
 function removeDoubleWhitespaces(prologProgram: string) {
+	const fetchDoubleWhitespacesRegExp = /  */gm;
+
 	return prologProgram.replace(fetchDoubleWhitespacesRegExp, " ");
 }
 
 export function removeNonFunctionalities(prologProgram: string) {
 	return removeComments(removeEmptyLines(removeDoubleWhitespaces(prologProgram)));
 }
+
+export function fixVariableShadowingInImport(program: string, importedProgram: string): string {
+	const fetchImportedPredicatesRegExp = /:-[\r\t\f\v ]*module\([^,)]*,\[([^)]*)\][\r\t\f\v ]*\)[\r\t\f\v ]*/;
+
+	const useModuleSecondParam: string[] | null = fetchImportedPredicatesRegExp.exec(importedProgram);
+	if (useModuleSecondParam == null) {
+		return identifyPredicatesUniquely(program, importedProgram);
+	}
+	// The following regex will return in group one the name of the functions
+	const importedPredicates: string[] = fetchAllMatchesForAGroup(useModuleSecondParam![1], /[\r\t\f\v ]*(\w*)\/\d,?/g, 1);
+	return identifyPredicatesUniquely(program, importedProgram, importedPredicates);
+}
+
+function identifyPredicatesUniquely(program: string, importedProgram: string, importedPredicates?: string[]): string {
+
+	// imports shouldn't be renamed
+	let predicatesToNotRename: string[];
+	// Checking for undefined based on this recommendation:
+	// https://basarat.gitbooks.io/typescript/docs/javascript/null-undefined.html
+	if (importedPredicates == null) {
```

Comment:
Just curious: Does this still work with `===` instead of `==`?

Reply:
It doesn't. Normally we would get an error for using `==` instead of `===`, but if comparing to null it is allowed without me needing to change anything. So the conclusion is `null` and `undefined` have the same value, yet are they different objects.

Reply:
Why don't you do `=== undefined` or `if(importedPredicates)` (checking the truthieness)?